src/dune_rules/melange/melange_stanzas.ml

```diff
@@ -51,14 +50,36 @@ module Emit = struct
     t
 
   let decode =
-    let extension_field name =
-      let+ loc, extension =
-        field name ~default:(Loc.none, "js") (located string)
-      in
+    let extension_field =
+      let+ loc, extension = located string in
       if String.is_prefix ~prefix:"." extension then
         User_error.raise ~loc [ Pp.textf "extension must not start with '.'" ];
       "." ^ extension
     in
+    let module_systems =
+      let module_system =
+        enum [ ("es6", Melange.Module_system.Es6); ("commonjs", CommonJs) ]
+      in
+      let+ module_systems =
+        repeat
+          (pair module_system (located extension_field)
+          <|> let+ loc, module_system = located module_system in
+              let _, ext = Melange.Module_system.default in
+              (module_system, (loc, ext)))
+      in
+      let (_ : String.Set.t), module_systems =
+        List.fold_map module_systems ~init:String.Set.empty
```

Comment:
If you were to use `String.Map.of_list_map`, you would be able to improve the error message by including both locations.